Add autodetection mode for add_docker_metadata

ChrsMark opened the pull request:
This PR is a proposal for enabling autodetection mode for `add_docker_metadata`.

Currently we configure `add_cloud_metadata` and `add_host_metadata` by default in all Beats.

`add_cloud_metadata` does autodetection, and it will only enrich events if the beat is running on a known public cloud.

It would be useful to have the same for Docker (and Kubernetes), so if Beats is started in a scenario where Docker is available, it automatically enables metadata for it. If Docker is not present nothing will happen.

In this PR only the Docker part is covered. After we conclude to a final approach, a follow-up PR will come for the Kubernetes part.

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -48,6 +52,18 @@ const (
+		logp.Info("%v: docker environment not detected.", processorName)
+		return
+	}
+	logp.Info("docker environment detected: %v", info)
```
I think this code should be called from `New`, as it is done in `add_cloud_metadata`, so the behaviour is that the processor is loaded, but it does nothing.

If we return here before the plugin is registered configurations will fail because `add_docker_metadata` will be an unknown processor.

jsoriano replied:
And then you wouldn't need the check when loading the list of processors :slightly_smiling_face:

ChrsMark commented:
@jsoriano thanks for reviewing!

Your suggestion has been addressed, however the checks are failing due to the addition in `filebeat.yml`. Any ideas on how I could resolve this?

jsoriano commented:
> Your suggestion has been addressed, however the checks are failing due to the addition in `filebeat.yml`. Any ideas on how I could resolve this?

`filebeat.yml` is generated from files in `libbeat/_meta`, you will need to make the change there and then run `make update` from the beats directory to regenerate all configuration files.

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -65,6 +69,18 @@ type addDockerMetadata struct {
 
 // New constructs a new add_docker_metadata processor.
 func New(cfg *common.Config) (processors.Processor, error) {
+	// check if Docker is available in running environment
```
Could you move the code to check if docker is available to a function? Then this comment will be probably unneeded :slightly_smiling_face:

ChrsMark replied:
Sure! Nice catch!

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -63,8 +67,30 @@ type addDockerMetadata struct {
+		logp.Info(errorMsg)
+		return fmt.Errorf(errorMsg)
+	}
+	logp.Info("docker environment detected: %v", info)
```
Nit. Add the processor name to this message too.
```suggestion
	logp.Infof("%v: docker environment detected: %v", processorName, info)
```

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -63,8 +67,30 @@ type addDockerMetadata struct {
 func New(cfg *common.Config) (processors.Processor, error) {
+	err := checkForDockerEnvironment()
+	if err != nil {
+		return nil, nil
```
Is it ok to return a `nil` processor here? Will something try to call `Run` over this nill object?

ChrsMark replied:
This is right. I should not return nil here. Thank you, will be addressed!

jsoriano commented on libbeat/_meta/config.yml.tmpl:
```diff
@@ -93,11 +93,12 @@ output.elasticsearch:
 processors:
   - add_host_metadata: ~
   - add_cloud_metadata: ~
+  - add_docker_metadata: ~
```
It could make sense to add it to `config.docker.yml` too.

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -57,10 +61,28 @@ type addDockerMetadata struct {
+	cli, err := client.NewEnvClient()
+	errorMsg := fmt.Sprintf("%v: docker environment not detected.", processorName)
+	if err != nil {
+		logp.Info(errorMsg)
```
Log this error at debug level, and with the error too?
```suggestion
		logp.Debugf("%s: %v", errorMsg, err)
```
And same thing for the other error some lines below.

ChrsMark commented:
Changed the implementation a little bit so as to perform the connectivity check for Docker with the proper configuration.
Also added a test for the case of Docker is unreachable.

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -19,6 +19,7 @@ package add_docker_metadata
 
 import (
 	"fmt"
+
```
Please remove this extra new line.
```suggestion
```

jsoriano commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -123,6 +132,9 @@ func lazyCgroupCacheInit(d *addDockerMetadata) {
 	var err error
+	if !d.dockerAvailable {
+		return event, nil
+	}
```
Nit. You could move this code to the very end of the function, before these declarations.

jsoriano commented on CHANGELOG.next.asciidoc:
```diff
@@ -244,6 +244,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - add_host_metadata is no GA. {pull}13148[13148]
 - Add `registered_domain` processor for deriving the registered domain from a given FQDN. {pull}13326[13326]
 - Add support for RFC3339 time zone offsets in JSON output. {pull}13227[13227]
+- Add autodetection mode for add_docker_metadata {pull}13374[13374]
```
Maybe we should mention something about enabling it by default.

```suggestion
- Add autodetection mode for add_docker_metadata and enable it by default in included configuration files {pull}13374[13374]
```

jsoriano commented on libbeat/common/docker/watcher.go:
```diff
@@ -132,6 +132,12 @@ func NewWatcher(host string, tls *TLSConfig, storeShortID bool) (Watcher, error)
+	_, err = client.Info(context.Background())
+	if err != nil {
+		return nil, err
+	}
```
I wonder if this can affect other users of `NewWatcher`, it seems safe though.

ChrsMark replied:
Yes it could affect others (`AutodiscoverBuilder`) in case of a connectivity error, but it wont be a problem I think, since an error will be returned for a case that is indeed problematic.
